### PR TITLE
Support Custom Secret Targets

### PR DESCRIPTION
**Pull request by ehazlett**
This adds support for custom secret target paths lifting the restriction of limiting secrets to `/var/run`.   Here is an example:

```
$> docker service create --name test --secret source=app,target=/etc/app -t alpine ash
# brace yourself for this one-liner...  yolo
$> docker exec -ti $(docker inspect --format '{{.Status.ContainerStatus.ContainerID}}' $(docker service ps --format '{{.ID}}' test)) mount | grep tmpfs | grep app
tmpfs on /etc/app type tmpfs (ro,relatime)
$> docker exec -ti $(docker inspect --format '{{.Status.ContainerStatus.ContainerID}}' $(docker service ps --format '{{.ID}}' test)) cat /etc/app
TESTING
```

Refs https://github.com/docker/docker/pull/32336#issuecomment-293035306
Depends on https://github.com/docker/swarmkit/pull/2118

### CODE REVIEW

File: vendor.conf [hunk 1]
```diff
@@ -105,7 +105,8 @@ github.com/docker/containerd 9048e5e50717ea4497b757314bad98ea3763c145
 github.com/tonistiigi/fifo 1405643975692217d6720f8b54aeee1bf2cd5cf4
 
 # cluster
-github.com/docker/swarmkit d5232280c510d70755ab11305d46a5704735371a
+github.com/docker/swarmkit d58bab69b2305019be600726bca5f82392f4f116 https://github.com/ehazlett/swarmkit
+github.com/golang/mock bd3c8e81be01eef76d4b503f5e687d2d1354d2d9
```
Comment (aaronlehmann) [inline vendor.conf]: Don't re-add mock. It was removed here: #32511

File: integration-cli/docker_cli_service_create_test.go [hunk 2]
```diff
@@ -90,35 +90,48 @@ func (s *DockerSwarmSuite) TestServiceCreateWithSecretSimple(c *check.C) {
...
+	d.DeleteSecret(c, testName)
+}
 
-	out, err = d.Cmd("service", "inspect", "--format", "{{ json .Spec.TaskTemplate.ContainerSpec.Secrets }}", serviceName)
-	c.Assert(err, checker.IsNil)
-
-	var refs []swarm.SecretReference
-	c.Assert(json.Unmarshal([]byte(out), &refs), checker.IsNil)
-	c.Assert(refs, checker.HasLen, 1)
+func (s *DockerSwarmSuite) TestServiceCreateWithSecretSourceTargetPaths(c *check.C) {
```
Comment (aaronlehmann) [inline integration-cli/docker_cli_service_create_test.go]: Slightly out of scope, but I realized recently that we don't seem to have tests that actually exec into the container to check that the secrets exist in the right place and have the right content. It's been on my list to add this. If you have the bandwidth to do it, I think it would be very valuable. Inspecting the container only goes so far - it doesn't confirm the secret mounts are working as expected.

File: container/container_unix.go [hunk 3]
```diff
@@ -253,29 +253,40 @@ func (container *Container) IpcMounts() []Mount {
...
+// SecretMounts returns the mount for the secret path
+func (container *Container) SecretMounts() []Mount {
+	var mounts []Mount
+	for _, r := range container.SecretReferences {
+		// secrets are created in the SecretMountPath at a single level
+		// i.e. /var/run/secrets/foo
+		srcPath := filepath.Join(container.SecretMountPath(), filepath.Base(r.File.Name))
+		mounts = append(mounts, Mount{
+			Source:      srcPath,
+			Destination: r.File.Name,
```
Comment (aaronlehmann) [inline container/container_unix.go]: Doesn't this need to be prefixed with `containerSecretMountPath` if the target is not an absolute path?

File: container/container_unix.go [hunk 4]
```diff
@@ -253,29 +253,40 @@ func (container *Container) IpcMounts() []Mount {
...
-		return &Mount{
-			Source:      container.SecretMountPath(),
-			Destination: containerSecretMountPath,
+// SecretMounts returns the mount for the secret path
+func (container *Container) SecretMounts() []Mount {
+	var mounts []Mount
+	for _, r := range container.SecretReferences {
+		// secrets are created in the SecretMountPath at a single level
+		// i.e. /var/run/secrets/foo
+		srcPath := filepath.Join(container.SecretMountPath(), filepath.Base(r.File.Name))
```
Comment (aaronlehmann) [inline container/container_unix.go]: What happens if two secret targets have the same base name?

File: container/container_unix.go [hunk 5]
```diff
@@ -253,29 +253,40 @@ func (container *Container) IpcMounts() []Mount {
...
+	return mounts
 }
 
 // UnmountSecrets unmounts the local tmpfs for secrets
 func (container *Container) UnmountSecrets() error {
-	if _, err := os.Stat(container.SecretMountPath()); err != nil {
-		if os.IsNotExist(err) {
-			return nil
+	for _, r := range container.SecretReferences {
+		srcPath := filepath.Join(container.SecretMountPath(), r.File.Name)
```
Comment (aaronlehmann) [inline container/container_unix.go]: In `SecretMounts` this uses the base name. The code here doesn't seem consistent.

Either way, we should break this out into a shared function.

File: integration-cli/docker_cli_service_create_test.go [hunk 6]
```diff
@@ -90,35 +90,48 @@ func (s *DockerSwarmSuite) TestServiceCreateWithSecretSimple(c *check.C) {
...
+	d := s.AddDaemon(c, true, true)
 
-	c.Assert(refs[0].SecretName, checker.Equals, testName)
-	c.Assert(refs[0].File, checker.Not(checker.IsNil))
-	c.Assert(refs[0].File.Name, checker.Equals, testTarget)
+	testPaths := map[string]string{
+		"app":         "/etc/secret",
+		"test_secret": "test_secret",
```
Comment (cpuguy83) [inline integration-cli/docker_cli_service_create_test.go]: Can we test directory traversal?

File: container/container_unix.go (hunk 5 above)
Comment (ehazlett) [inline container/container_unix.go]: Good point :+1:

File: integration-cli/docker_cli_service_create_test.go (hunk 2 above)
Comment (ehazlett) [inline integration-cli/docker_cli_service_create_test.go]: Same here.  I'll get some added.

**Comment by thaJeztah** [comment]
Design looks good to me 👍 Moving this to code review, but let me know if there's objections 😄

**Comment by aaronlehmann** [comment]
It still looks like this will break when two secrets have the same target base name.

**Comment by aaronlehmann** [comment]
@ehazlett: I'd suggest this borrow the code from configs in #32336. That stores the config under the full target path, so there isn't a risk of collisions from targets that share the same basename. Also, it would be good for the config and secret implementations to stay consistent.

Let me know if it would be helpful for me to carry this PR.

**Comment by ehazlett** [comment]
Ya if you want to carry that's fine with me.

**Comment by aaronlehmann** [comment]
I've updated the vendoring, as the upstream swarmkit PR was merged.

I added test coverage for the following:

- Verify the presence and contents of secrets with simple, relative, and absolute targets
- Ensure that directory traversal is prevented
- Check that a service can reference the same secret multiple times under different targets

I also added corresponding tests to the configs PR (#32336).

PTAL

**Comment by thaJeztah** [comment]
Thanks @aaronlehmann!

ping @cyli @riyazdf ptal

**Comment by aaronlehmann** [comment]
`TestSwarmLockUnlockCluster` appears to have hung on `powerpc` and `z`. I'd guess this is related to the swarmkit vendoring rather than the specific changes in this PR. We're working on getting an etcd/raft issue fixed upstream - this might be related.

File: container/container.go [hunk 7]
```diff
@@ -949,3 +949,26 @@ func (container *Container) InitializeStdio(iop libcontainerd.IOPipe) error {
...
+
+// SecretMountPath returns the path of the secret mount for the container
+func (container *Container) SecretMountPath() string {
+	return filepath.Join(container.Root, "secrets")
+}
+
+// SecretFilePath returns the path to the location of a secret on the host.
+func (container *Container) SecretFilePath(secretRefTarget *swarmtypes.SecretReferenceFileTarget) (string, error) {
+	secretMountPath := container.SecretMountPath()
+	mountPath := filepath.Join(secretMountPath, secretRefTarget.Name)
```
Comment (diogomonica) [inline container/container.go]: Are we ok with joining `secretRefTarget.Name` directly? I'm assuming our regex doesn't allow any path separator characters in names? (it definitely allows `.`)

File: container/container_unix.go [hunk 8]
```diff
@@ -253,29 +248,52 @@ func (container *Container) IpcMounts() []Mount {
...
 }
 
-// SecretMount returns the mount for the secret path
-func (container *Container) SecretMount() *Mount {
-	if len(container.SecretReferences) > 0 {
-		return &Mount{
-			Source:      container.SecretMountPath(),
-			Destination: containerSecretMountPath,
-			Writable:    false,
+// SecretMounts returns the mount for the secret path
```
Comment (diogomonica) [inline container/container_unix.go]: *mounts

File: integration-cli/docker_cli_service_create_test.go [hunk 9]
```diff
@@ -90,35 +91,128 @@ func (s *DockerSwarmSuite) TestServiceCreateWithSecretSimple(c *check.C) {
...
 
-func (s *DockerSwarmSuite) TestServiceCreateWithSecretSourceTarget(c *check.C) {
+func (s *DockerSwarmSuite) TestServiceCreateWithSecretSourceTargetPaths(c *check.C) {
+	d := s.AddDaemon(c, true, true)
+
+	testPaths := map[string]string{
+		"app":             "/etc/secret",
+		"test_secret":     "test_secret",
+		"relative_secret": "relative/secret",
+		"invalid_secret":  "../secret",
```
Comment (diogomonica) [inline integration-cli/docker_cli_service_create_test.go]: Should we even allow a secret create w/ a target like this?

File: opts/secret.go [hunk 10]
```diff
@@ -53,10 +52,6 @@ func (o *SecretOpt) Set(value string) error {
 		case "source", "src":
 			options.SecretName = value
 		case "target":
-			tDir, _ := filepath.Split(value)
```
Comment (diogomonica) [inline opts/secret.go]: Why are we fine with a target being a directory? This still doesn't make sense right?

**Review by diogomonica — COMMENTED** [review]
Couple of questions:
- Should we even support secret creation with targets that have potential OS separators?
- Should we list all secrets in `/var/run/secrets` even if the target is outside of base secrets dir? Would have the advantage that a `ls /var/run/secrets/` always returns all the secrets available.
- Should we allow someone to add a target into `/var/run/secrets/`? In particular concerned that the validation of multiple targets being set somehow doesn't handle this new way of adding targets and the behavior will be unknown or the task unschedulable.
- Are there any files/paths we want to reject outright from being overwritten? (like anything in `/proc/`?)

Thanks for carrying this PR @aaronlehmann

File: container/container.go (hunk 7 above)
Comment (aaronlehmann) [inline container/container.go]: It does allow path separators - that's the point of https://github.com/docker/swarmkit/pull/2118.

The check below tries to ensure that this doesn't escape the mount point, but I'd appreciate careful review to make sure this works as intended.

File: integration-cli/docker_cli_service_create_test.go (hunk 9 above)
Comment (aaronlehmann) [inline integration-cli/docker_cli_service_create_test.go]: The way paths work is OS-specific, so deciding in the manager or the client whether a path is acceptable is a bit tricky. We don't know what OS the service will eventually run on.

File: opts/secret.go (hunk 10 above)
Comment (aaronlehmann) [inline opts/secret.go]: The client code has no way of knowing whether a path is a file or a directory.

This old code was essentially checking that there was no path separator. It is being removed because now path separators are allowed.

**Comment by aaronlehmann** [comment]
> Should we even support secret creation with targets that have potential OS separators?

This one I don't understand. The purpose of this PR is to allow you to put secrets anywhere on the filesystem, not just `/run/secrets`. For this to work, it must be possible to include path separators in the target. Do you mean something different by "OS separators"?

> Should we list all secrets in /var/run/secrets even if the target is outside of base secrets dir? Would have the advantage that a ls /var/run/secrets/ always returns all the secrets available.

I'm not sure how it would work. What would we use as the filename under `/run/secrets` that would be guaranteed to be unique? I suppose we could use secret ID but it would not be very user-friendly. Or we could put a secret with the target `/path/to/target` in both `/path/to/target` and `/run/secrets/path/to/target`, but I feel like that might be confusing.

> Should we allow someone to add a target into /var/run/secrets/? In particular concerned that the validation of multiple targets being set somehow doesn't handle this new way of adding targets and the behavior will be unknown or the task unschedulable.

Let's make sure no two secrets have the same `getSecretTargetPath` result.

I'm okay with also disallowing targets that explicitly use a path under `/run/secrets`, but the above would be sufficient to prevent conflicts.

> Are there any files/paths we want to reject outright from being overwritten? (like anything in /proc/?)

Since we use bind mounts, I don't think we're actually overwriting anything. I can't think of any paths that would be important to block, but let me know if you think of any.

**Comment by aaronlehmann** [comment]
@diogomonica: I checked, and putting a target in `/run/secrets` in such a way that it conflicts with an implicit path is already prevented:

```
ERRO[0125] fatal task error                              error="starting container failed: linux mounts: Duplicate mount point '/run/secrets/foo'" module="node/agent/taskmanager" node.id=eztzkyt8y44569gywcgl7t03a service.id=02usncjhurcy4di7q5wrv3ij2 task.id=qvrp1dwru2ittc7dhg560xluy
```

**Comment by diogomonica** [comment]
LGTM

File: container/container_windows.go [hunk 11]
```diff
@@ -43,8 +47,8 @@ func (container *Container) IpcMounts() []Mount {
 	return nil
 }
 
-// SecretMount returns the mount for the secret path
-func (container *Container) SecretMount() *Mount {
+// SecretMounts returns the mount for the secret path
```
Comment (cyli) [inline container/container_windows.go]: nitpick:  "mounts" here too :)

File: daemon/container_operations_unix.go [hunk 12]
```diff
@@ -177,13 +177,15 @@ func (daemon *Daemon) setupSecretDir(c *container.Container) (setupErr error) {
...
-		targetPath := filepath.Clean(s.File.Name)
-		// ensure that the target is a filename only; no paths allowed
-		if targetPath != filepath.Base(targetPath) {
-			return fmt.Errorf("error creating secret: secret must not be a path")
-		}
+		log := logrus.WithField("name", s.File.Name)
 
-		fPath := filepath.Join(localMountPath, targetPath)
+		// secrets are created in the SecretMountPath at a single level
+		// i.e. /var/run/secrets/foo
```
Comment (cyli) [inline daemon/container_operations_unix.go]: Non-blocking comment nit:  Should this be something like `/var/lib/docker/devicemapper/mnt/<container sha>/secrets/foo` (even though they may not be using devicemapper, just to make it clear that `SecretMountPath` is on the host and not in a container?

**Comment by aaronlehmann** [comment]
Fixed nits

File: integration-cli/docker_cli_service_create_test.go (hunk 9 above)
Comment (cyli) [inline integration-cli/docker_cli_service_create_test.go]: Would checking `path == filepath.Clean(path)` work across platform, and just clean any references to `..`?  It might be more explicit to fail this way.

If I understand correctly, this particular case is invalid because on the host it would take you out of the flat mount directory, but something like `ignoreddir/../test_secret` would be fine because that would end up being just `secret`.  That might be confusing to users, since they may just want it in `/var/run/test_secret` as opposed to `/var/run/secrets/test_secret`, and may not understand that the `../` is disallowed because it's taking them out of the host path specifically.

Comment (aaronlehmann) [inline integration-cli/docker_cli_service_create_test.go]: I don't think it would work across platforms, because there are different implementations for Windows and Linux. Even if it does work, I wouldn't be super comfortable with passing Windows paths to `filepath.Clean` on a Linux host and vice versa.

I suppose we could grep for `(^|/)\.\./` and `(^|\)\.\.\`. This would take a PR to swarmkit, though.

Comment (cyli) [inline integration-cli/docker_cli_service_create_test.go]: Would it be feasible to have the filename be the secret ID or something else that wouldn't be traversable in the host directory?  That way we wouldn't have to be afraid of traversals out of the container root.

Comment (cpuguy83) [inline integration-cli/docker_cli_service_create_test.go]: We can at least let the worker evaluate the path and reject it. Also see: https://github.com/moby/moby/blob/master/pkg/symlink/fs.go#L21

**Comment by aaronlehmann** [comment]
I've pushed an update that uses the secret ID instead of the secret target as the filename for storing the secret on the host's tmpfs mount.

**Comment by cyli** [comment]
Thanks @aaronlehmann!  I could not think of any issues allowing a secret to escape `/run/secrets` inside a container, since we already allow absolute paths inside the container, whomever provides the target can get to any path anyway.  

I don't think users should be providing traversals from `/run/secrets` rather than absolute paths, but other than them confusing themselves, I don't think it presents any vulnerability if we're just using IDs (which we control) on the host.

Thinking about it more, it may be nice if we could prevent path traversals at all on `service create` but since as you mentioned that could be tricky, I'm not sure it's worth the effort, but would appreciate others giving this line of reasoning a sanity check :)

LGTM on green.

**Comment by diogomonica** [comment]
@cyli @aaronlehmann since we don't see any security issues and this would just be a hygienic decision (no transversal), it would probably just be best effort.

I also agree that we shouldn't allow relative paths at all. Unless someone can come up with a good use case.

**Comment by aaronlehmann** [comment]
Rebased

**Comment by aaronlehmann** [comment]
Corresponding CLI PR in https://github.com/docker/cli/pull/50 (really just vendoring `opts`, which is a bit odd...)

**Review by johnstep — APPROVED** [review]
LGTM

**Comment by cpuguy83** [comment]
```
Handler for DELETE /v1.30/containers/9e returned error: Unable to remove filesystem for 9e6c79a949b080473b9d4778f6eb9b3c9fb06a142260912457d57306a3e6d476: remove /var/lib/docker/containers/9e6c79a949b080473b9d4778f6eb9b3c9fb06a142260912457d57306a3e6d476/secrets: device or resource busy
```

This seems to happen with any secret target right now.

And I end up with leaked mounts:
```
tmpfs on /var/lib/docker/containers/67c604df32844744fea5d2613dc850d1c85227fc0a9406f4008938b4975358e4/secrets type tmpfs (rw,nosuid,nodev,noexec,relatime)
```

File: container/container_unix.go [hunk 13]
```diff
@@ -253,29 +248,43 @@ func (container *Container) IpcMounts() []Mount {
...
+				return nil
+			}
+			return err
+		}
+
+		if err := detachMounted(srcPath); err != nil {
+			return err
 		}
-		return err
 	}
```
Comment (cpuguy83) [inline container/container_unix.go]: I think after this is done we need to `detachMounted(container.SecretMountPath())`

Comment (aaronlehmann) [inline container/container_unix.go]: You are right. I'm not sure why this got removed. I think we should `defer` the original contents of this function - do you agree?

Comment (cpuguy83) [inline container/container_unix.go]: Yep, makes sense.

Comment (aaronlehmann) [inline container/container_unix.go]: @cpuguy83: I believe I've fixed this. I'm not seeing any errors or leaked mounts when I delete a container that uses secrets. I'd appreciate it if you could confirm on your side.

File: container/container_unix.go [hunk 14]
```diff
@@ -253,29 +248,59 @@ func (container *Container) IpcMounts() []Mount {
...
-	return nil
+	return mounts
 }
 
 // UnmountSecrets unmounts the local tmpfs for secrets
-func (container *Container) UnmountSecrets() error {
-	if _, err := os.Stat(container.SecretMountPath()); err != nil {
-		if os.IsNotExist(err) {
-			return nil
+func (container *Container) UnmountSecrets() (err error) {
```
Comment (cpuguy83) [inline container/container_unix.go]: Somewhere in here we are now getting a spurious EINVAL.
Maybe we can just check `if err == unix.EINVAL { if mount.Mounted(path) { return err }}`

Although I'm really curious as to where it's happening.

File: container/container_unix.go [hunk 15]
```diff
@@ -253,29 +248,59 @@ func (container *Container) IpcMounts() []Mount {
...
+	}()
+
+	for _, r := range container.SecretReferences {
+		if r.File == nil {
+			continue
+		}
+		srcPath := container.SecretFilePath(*r)
+		if _, err := os.Stat(srcPath); err != nil {
+			if os.IsNotExist(err) {
+				return nil
```
Comment (cpuguy83) [inline container/container_unix.go]: continue

File: container/container_unix.go (hunk 14 above)
Comment (cpuguy83) [inline container/container_unix.go]: I don't think we should return errors anywhere in this function since it stops all processing but cleanup should still happen for any remaining items.
The parent is only logging this error anyway.

Maybe if we do want to continue to return errors we can just collect all the errors into a single error.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: container/container_unix.go (hunk 14 above)
Comment (ehazlett) [inline container/container_unix.go]: For a little more context:
```
DEBU[3206] containerd: process exited                    id=2759b664f3a609398c03bcbc6af6ee3e9996097c2363d63a7c5a0100cf79037d pid=init status=0 systemPid=1631
DEBU[3206] libcontainerd: received containerd event: &types.Event{Type:"exit", Id:"2759b664f3a609398c03bcbc6af6ee3e9996097c2363d63a7c5a0100cf79037d", Status:0x0, Pid:"init", Timestamp:(*timestamp.Timestamp)(0xc421526db0)} 
DEBU[3206] Revoking external connectivity on endpoint test.1.3g1qqhu9spa9apbz7t28ixkvv (a8d45de538bfa41fe4b92dc866eda1426a653ebf79e68480ee768905c96f764d) 
DEBU[3206] DeleteConntrackEntries purged ipv4:0, ipv6:0 
DEBU[3206] Releasing addresses for endpoint test.1.3g1qqhu9spa9apbz7t28ixkvv's interface on network bridge 
DEBU[3206] ReleaseAddress(LocalDefault/172.18.0.0/16, 172.18.0.2) 
WARN[3206] 2759b664f3a609398c03bcbc6af6ee3e9996097c2363d63a7c5a0100cf79037d cleanup: failed to unmount secrets: invalid argument 
```

**Comment by aaronlehmann** [comment]
I've reverted all changes to `UnmountSecrets`. I think these were incorrect to begin with. The individual secret mounts are passed on to `runc` and should be cleaned up by `runc`. All that's necessary is unmounting the tmpfs on the host.

PTAL

File: hack/dockerfile/binaries-commits [hunk 16]
```diff
@@ -11,5 +11,5 @@ VNDR_COMMIT=c56e082291115e369f77601f9c071dd0b87c7120
 BINDATA_COMMIT=a0ff2567cfb70903282db057e799fd826784d41d
 
 # CLI
-DOCKERCLI_REPO=https://github.com/docker/cli
-DOCKERCLI_COMMIT=c3648a9c9400d45524cc71b8fca4085b192c626f
+DOCKERCLI_REPO=https://github.com/aaronlehmann/cli
+DOCKERCLI_COMMIT=fd5a5910409fe5ed862b29de45a66f2bf8056894
```
Comment (cpuguy83) [inline hack/dockerfile/binaries-commits]: Need to change these back before merge.

**Review by cpuguy83 — APPROVED** [review]
LGTM

File: hack/dockerfile/binaries-commits (hunk 16 above)
Comment (aaronlehmann) [inline hack/dockerfile/binaries-commits]: They can't be changed back before the merge. This would cause CI to fail - unless you think the CLI PR should be merged before this one (but that seems backwards).

These circular dependencies are very hard to deal with.

Comment (johnstep) [inline hack/dockerfile/binaries-commits]: Don't we need to keep this until docker/cli#50 is merged?

Comment (cpuguy83) [inline hack/dockerfile/binaries-commits]: We can move the tests to a separate PR.

Opts package should really not be in this repo.
We're only using a couple of functions from it for cmd/dockerd.

Comment (aaronlehmann) [inline hack/dockerfile/binaries-commits]: The plan I discussed with @tiborvass and @thaJeztah was to merge this PR with the `DOCKERCLI_REPO` change intact, then change this back to the main repo in a followup.

Comment (cpuguy83) [inline hack/dockerfile/binaries-commits]: Really even the opts change could be made on it's own since the existing code is not valid as is.

Comment (cpuguy83) [inline hack/dockerfile/binaries-commits]: @aaronlehmann That's fine.

**Comment by ehazlett** [comment]
FWIW tested on my end again and LGTM
